File: src/ir/primitive/ast-identifier.ts
import type { AST_TYPE } from "../../parser/ast-type.js";
import type { AstNode } from "./ast-node.js";

export type AstIdentifier = AstNode &
  Readonly<{
    type: typeof AST_TYPE.Identifier | typeof AST_TYPE.JSXIdentifier;
    name: string;
  }>;

File: crates/unsnarl-ir/src/primitive/ast_identifier.rs
//! Materialised Identifier / JSXIdentifier node.

use oxc_span::Span;

use unsnarl_oxc_parity::AstType;

use crate::non_empty::assert_non_empty;

#[derive(Clone)]
pub struct AstIdentifier {
    pub r#type: AstType,
    name: String,
    pub span: Span,
}

impl AstIdentifier {
    pub fn new(r#type: AstType, name: String, span: Span) -> Self {
        assert_non_empty(&name, "AstIdentifier.name");
        Self { r#type, name, span }
    }

    pub fn name(&self) -> &str {
        &self.name
    }
}
